### ISSUE TITLE
Extra spaces inserted into certificates

### ISSUE THREAD

**Issue by evilhamsterman** [vmware/vic-product#2173]
Please fill out both the Basic Information and Detailed Information sections.

Bug Report Basic Information
----------------------------

REQUIRED:

- vCenter Server version: 6.7U1
- Embedded or external PSC: Embedded
- Filename of the OVA you deployed: vic-v1.4.4-6375-0a6da3d4.ova
- How was the OVA deployed? (Flex client, HTML5 client, ovftool): HTML5
- Does the VIC appliance receive configuration by DHCP?: yes
- What stage of the Appliance Lifecycle is the VIC appliance in?: Running
  https://github.com/vmware/vic-product/blob/master/installer/docs/SUPPORT.md#appliance-lifecycle
- Attach the VIC appliance support bundle (e.g. vic_appliance_logs_2018-01-01-00-01-00.tar.gz)[vic_appliance_logs_2018-11-09-21-43-41.tar.gz](https://github.com/vmware/vic-product/files/2567606/vic_appliance_logs_2018-11-09-21-43-41.tar.gz)

  https://github.com/vmware/vic-product/blob/master/installer/docs/SUPPORT.md#appliance-support-bundle
  Note: The support bundle may contain private information. If you are not comfortable with posting this publicly, please contact VMware GSS.

OPTIONAL, but helpful:

- IP address of VIC appliance: 10.220.248.57
- Hostname of VIC appliance: vic-poc.corp.qumulo.com
- IP address of vCenter Server: 10.220.247.75
- Hostname of vCenter Server: vc-poc.corp.qumulo.com

After deploying the VIC appliance with custom internally signed certs Admiral is unable to start. Upon digging into the appliance the Admiral start up is complaining it cannot read either the key or the certificates. Opening the key and certificate files shows extra spaces have been added into the files. Other services ignore the spaces but the Admiral service is unable to properly read the files. I was able to resolve it not being able to read the key by providing a PKCS#1 formatted key so it was auto converted so no spaces were added. 

The problem does not occur when using the Flash client.

The keys and certificates all appear similar to this
```
-----BEGIN CERTIFICATE-----
 MIIGADCCBOigAwIBAgIKaw3HyAABAAAAODANBgkqhkiG9w0BAQsFADBXMRMwEQY
K CZImiZPyLGQBGRYDY29tMRYwFAYKCZImiZPyLGQBGRYGcXVtdWxvMRIwEAYKCZ
Im iZPyLGQBGRYCYWQxFDASBgNVBAMTC2FkLVNFQTAxLUNBMB4XDTE4MTEwOTE4M
DUz MVoXDTIwMTEwODE4MDUzMVowdDELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldh
c2hp bmd0b24xEDAOBgNVBAcTB1NlYXR0bGUxDzANBgNVBAoTBlF1bXVsbzELMAk
GA1UE CxMCSVQxIDAeBgNVBAMTF3ZpYy1wb2MuY29ycC5xdW11bG8uY29tMIIBIj
ANBgkq hkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAt31DyOtxrMQbtSvn14vukBdK+
abqV/74 sV0vtIIJO/ecZX1gxSSoMeOpoa5itwczy1zH4BK8w3APA1RbRyQ8FA42
Iljuj+k/ s2yZNcnNgX521qmhiaMFXpVor+LJsZ2D74x5P6vBLebxWglgnc5IHop
nMiL4KhDI ft2Nl332ADoDeMG8jap/OZFOnUu5Fo5OYIybhCCoZbf9pVP/4KUJrh
LpSFE3/umq dZAC6PzEpLeEHQOK7AL+SabyRUr9syQjIdwu9wz81zos75fucsRsV
D/Kt+aScxcf c3dghWo1E5j3fxm2wVN/b9/gcQuSard+bcuSOvmXyU9N+qMWYA9B
IQIDAQABo4IC rzCCAqswDgYDVR0PAQH/BAQDAgXgMB0GA1UdJQQWMBQGCCsGAQU
FBwMBBggrBgEF BQcDAjAxBgNVHREEKjAohwQK3Pg5ghd2aWMtcG9jLmNvcnAucX
VtdWxvLmNvbYIH dmljLXBvYzAdBgNVHQ4EFgQUsxGFPFw8XkY+ORHMRV4oIQnew
h4wHwYDVR0jBBgw FoAUA05ucZOp3Bd1nivZRRh1fZ7IsWQwggEBBgNVHR8Egfkw
gfYwgfOggfCgge2G gbhsZGFwOi8vL0NOPWFkLVNFQTAxLUNBKDEpLENOPXNlYTA
xLENOPUNEUCxDTj1Q dWJsaWMlMjBLZXklMjBTZXJ2aWNlcyxDTj1TZXJ2aWNlcy
xDTj1Db25maWd1cmF0 aW9uLERDPWFkLERDPXF1bXVsbyxEQz1jb20/Y2VydGlma
WNhdGVSZXZvY2F0aW9u TGlzdD9iYXNlP29iamVjdENsYXNzPWNSTERpc3RyaWJ1
dGlvblBvaW50hjBodHRw Oi8vY3JsLmFkLnF1bXVsby5jb20vY3JsZC9hZC1TRUE
wMS1DQSgxKS5jcmwwgcIG CCsGAQUFBwEBBIG1MIGyMIGvBggrBgEFBQcwAoaBom
xkYXA6Ly8vQ049YWQtU0VB MDEtQ0EsQ049QUlBLENOPVB1YmxpYyUyMEtleSUyM
FNlcnZpY2VzLENOPVNlcnZp Y2VzLENOPUNvbmZpZ3VyYXRpb24sREM9YWQsREM9
cXVtdWxvLERDPWNvbT9jQUNl cnRpZmljYXRlP2Jhc2U/b2JqZWN0Q2xhc3M9Y2V
ydGlmaWNhdGlvbkF1dGhvcml0 eTA9BgkrBgEEAYI3FQcEMDAuBiYrBgEEAYI3FQ
iHuOwv0pZdhbWbAYX7ih+F+ct2 gWqHpc4mgcKzWgIBZAIBAjANBgkqhkiG9w0BA
QsFAAOCAQEAE/jiBr2nrpdMhkCW YLayA6OWPe1bIZ3yVqyIxyu7TaVXUzlew+GY
9zi+9Bf5pm8JtcdmO8er5IBqrFad HHb+QgbHWExEWtqV1HPi0iSR7WPqV5NG6+r
OnzdOi5jMpWsHfPWxsYn0in1eS9nL KaPSZfkMwGCLA0fP45SoGKzRKUu4ajqqGX
9dhCL+dYufBmK2o3IRWhdQGtqYUboQ goFxV0sxv3ofQfhY5/A1zKsRKH1kuvWxr
73iam85DpiqWOoMb7bKHxivQN5p4bwA Juk/rAt0anq4+OB3RbqDQA1IDxguICOz
XGtkrYrey6MJ+djYKTv1LFRMp4HWaGpd 3CrgSg==
-----END CERTIFICATE-----
 -----BEGIN CERTIFICATE-----
 MIIDsDCCApigAwIBAgIQW615hyCROIRKyqOyhmCabTANBgkqhkiG9w0BAQsFADB
X MRMwEQYKCZImiZPyLGQBGRYDY29tMRYwFAYKCZImiZPyLGQBGRYGcXVtdWxvMR
Iw EAYKCZImiZPyLGQBGRYCYWQxFDASBgNVBAMTC2FkLVNFQTAxLUNBMB4XDTE4M
DQw MjE5MDU1MFoXDTI4MDQwMTE5MTU0OVowVzETMBEGCgmSJomT8ixkARkWA2Nv
bTEW MBQGCgmSJomT8ixkARkWBnF1bXVsbzESMBAGCgmSJomT8ixkARkWAmFkMRQ
wEgYD VQQDEwthZC1TRUEwMS1DQTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQ
oCggEB ALS0TpyAkrjOl7vpiVpl8EdB7MQaiSbAZ7GdmNtuTvBLUU+PBzjISlW2R
sKfne1e i9D8+BclgyxaHVtRjhzU4+r4ZT2a4MXySiJPH2hu4jPgXxYAaEhRNd3G
SBS7OgVN GBf9CrOLdkA9ht7hfZDQwbZXzXNYrttkkSy9+5zlgrbb822jMsgaGiI
FjR247xiW cvrkRPDiR+wDuKRkV3TLVXwNERK1nu/1rNXMjnE9JnwBsPMOTVUWdM
4tsh2cSait YYLZtou8gWsCckhRTQObv2itx1Q0jJWsj1DfoW3cS3k78OQ2ike91
mZXd2X6pMyD s3g++a38hEfn8oB3c98vORECAwEAAaN4MHYwCwYDVR0PBAQDAgGG
MA8GA1UdEwEB /wQFMAMBAf8wHQYDVR0OBBYEFANObnGTqdwXdZ4r2UUYdX2eyLF
kMBIGCSsGAQQB gjcVAQQFAgMBAAEwIwYJKwYBBAGCNxUCBBYEFM52O2cZLSKgqx
cLhMebxFSCLqub MA0GCSqGSIb3DQEBCwUAA4IBAQBRW1hvi/9Az+hmVkl2O5uPj
k20GtDIt3VQ/223 m4W5HqYOsx5b0Iy/qp+JublPXk/OG3hmaTlMzcefuSwv0PAj
69gwSIcCwQjxivDo 22hYgq2vVLq4ZKYlie6m1UrT66jcoKIRtyc7BugFJm78FBU
S8rf4aBYH3Q1Hpo+Q OIG4m7nUoGdfCY1YFCx8cRBNXXj1iAclC5CDlUgD3P6qI/
8i2BV6HZHgwEzd34SE HfQl+FYR8Lul+pnHlYFarpQRngLbdw7PjnQj3DrvADcGR
p5FDpwof4/HKJqCp6cX MABeWL7EKSFkpSffhjtXI847RPMaPWyjFSFmEq9dyVlM
8IoZ
-----END CERTIFICATE-----
```

**Comment by wjun** [comment]
@evilhamsterman Just double confirm that the cert files you provided do not have white spaces as shown above, and when you deploy the ova through HTML5 web client, the generated vic appliance added white spaces into cert files?

**Comment by evilhamsterman** [comment]
Yes they are valid certs with none of the whitespaces show. When I deploy using the HTML5 client they are added, they are not added when I use the Flash client.

**Comment by evilhamsterman** [comment]
As an update I think I may have found part of the problem. The certificate was issued by our Active Directory CA and exported on a Windows system so the certificate had Windows line endings (\r\n vs just \n). It seems the Flash interface strips all the line endings but the HTML5 one does not and whatever scripts your using to process the certificates somehow convert he extra \r into whitespace. I cleared the \r from my files which fixes the problems but your tools should probably clear it or at least you should document to ensure the certs only have \n endings

**Comment by evilhamsterman** [comment]
Disregard I tried again with cert files I ran through the `dos2unix` command and got the same result. Extra whitespaces added to the certificates this time causing my upgrade from 1.4.4 to 1.5.1 to fail

**Comment by luwang-vmware** [comment]
The reason is that there is space in every 64 char in the certs/keys content which shows in appliance vApp options. 

If there shows any space in the certs/key content, then Change the L43 in /etc/vmware/vic-appliance-tls.sh with   ****echo "$content" | sed -r 's/(-{5}BEGIN [A-Z ]+-{5})/&\n/g; s/(-{5}END [A-Z ]+-{5})/\n&\n/g' | sed 2's/ //g' | sed -r 's/.{64}/&\n/g; /^\s*$/d' | sed -r     '/^$/d' > "$file**"** to replace the former one    echo "$content" | sed -r 's/(-{5}BEGIN [A-Z ]+-{5})/&\n/g; s/(-{5}END [A-Z ]+-{5})/\n&\n/g' | sed -r 's/.{64}/&\n/g; /^\s*$/d' | sed -r     '/^$/d' > "$file". And reboot the appliance to take it effect.

**Comment by cmbits** [comment]
I ran into the same issue with vCenter 6.7.0.1150 build 9433921 and VIC version 1.5.2-7206.  I tried 20 deployments of the ova using different certificate formats through the HTML5 vcenter interface and I would typically get the following error; "java.lang.IllegalArgumentException: Illegal base64 character 20".  After reading this I tried the vCenter flex interface and the deployment worked just fine.  Thank you evilhampsterman.  You saved me a few more days of failed attempts.
